File: fixtures/lang-coverage/objc/src/Service.m
// feature: @implementation, method definition, call (local + message send)
#import "Service.h"

@implementation Service

- (instancetype)initWithName:(NSString *)name {
    self = [super init];
    if (self) {
        _name = name;
        _state = ServiceStateIdle;
    }
    return self;
}

- (void)run {
    NSString *greeting = [Helpers formatName:self.name];
    NSLog(@"%@", greeting);
    _state = ServiceStateRunning;
}

- (NSString *)greeting {
    return [NSString stringWithFormat:@"hello from %@", _name];
}

+ (id)shared {
    static Service *instance = nil;
    if (!instance) {
        instance = [[Service alloc] initWithName:@"default"];
    }
    return instance;
}

@end

File: fixtures/lang-coverage/pascal/src/service.pas
unit Service;

interface

uses SysUtils, Types, Helpers;

type
  TServiceRunner = class
  private
    FName: string;
  public
    constructor Create(const Name: string);
    procedure Run;
    function GetGreeting: string;
  end;

implementation

constructor TServiceRunner.Create(const Name: string);
begin
  FName := Name;
end;

procedure TServiceRunner.Run;
var
  Greeting: string;
begin
  Greeting := FormatName(FName);
  WriteLn(Greeting);
end;

function TServiceRunner.GetGreeting: string;
begin
  Result := FormatName(FName);
end;

end.
